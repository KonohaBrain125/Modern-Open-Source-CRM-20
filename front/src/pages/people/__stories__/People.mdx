{ /* People.mdx */ }

import { Canvas, Meta } from '@storybook/blocks';

import * as People from './People.stories';

<Meta of={People} />

# People View

<Canvas of={People.Default} />