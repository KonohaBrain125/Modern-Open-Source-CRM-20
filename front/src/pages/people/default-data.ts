import { GraphqlQueryPerson } from '../../interfaces/person.interface';

export const defaultData: Array<GraphqlQueryPerson> = [
  {
    id: '7dfbc3f7-6e5e-4128-957e-8d86808cdf6b',
    __typename: 'Person',
    firstname: 'Alexandre',
    lastname: 'Prot',
    email: 'alexandre@qonto.com',
    company: {
      id: '7dfbc3f7-6e5e-4128-957e-8d86808cdf6c',
      name: 'Qonto',
      domain_name: 'qonto.com',
      __typename: 'Company',
    },
    phone: '06 12 34 56 78',
    created_at: '2023-04-20T13:20:09.158312+00:00',

    city: 'Paris',
  },
  {
    id: '7dfbc3f7-6e5e-4128-957e-8d86808cdf6d',
    __typename: 'Person',
    firstname: 'John',
    lastname: 'Doe',
    email: 'john@linkedin.com',
    company: {
      id: '7dfbc3f7-6e5e-4128-957e-8d86808cdf6e',
      name: 'LinkedIn',
      domain_name: 'linkedin.com',
      __typename: 'Company',
    },
    phone: '06 12 34 56 78',
    created_at: '2023-04-20T13:20:09.158312+00:00',

    city: 'Paris',
  },
  {
    id: '7dfbc3f7-6e5e-4128-957e-8d86808cdf6f',
    __typename: 'Person',
    firstname: 'Jane',
    lastname: 'Doe',
    email: 'jane@sequoiacap.com',
    company: {
      id: '7dfbc3f7-6e5e-4128-957e-8d86808cdf6g',
      name: 'Sequoia',
      domain_name: 'sequoiacap.com',
      __typename: 'Company',
    },
    phone: '06 12 34 56 78',
    created_at: '2023-04-20T13:20:09.158312+00:00',

    city: 'Paris',
  },

  {
    id: '7dfbc3f7-6e5e-4128-957e-8d86808cdf6h',
    __typename: 'Person',
    firstname: 'Janice',
    lastname: 'Dane',
    email: 'janice@facebook.com',
    company: {
      id: '7dfbc3f7-6e5e-4128-957e-8d86808cdf6i',
      name: 'Facebook',
      domain_name: 'facebook.com',
      __typename: 'Company',
    },
    phone: '06 12 34 56 78',
    created_at: '2023-04-20T13:20:09.158312+00:00',

    city: 'Paris',
  },
];
