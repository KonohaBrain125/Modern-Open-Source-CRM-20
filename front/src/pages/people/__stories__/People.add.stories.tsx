import { expect } from '@storybook/jest';
import type { Meta } from '@storybook/react';
import { userEvent, within } from '@storybook/testing-library';

import { AppPath } from '@/types/AppPath';
import {
  PageDecorator,
  type PageDecoratorArgs,
} from '~/testing/decorators/PageDecorator';
import { graphqlMocks } from '~/testing/graphqlMocks';
import { sleep } from '~/testing/sleep';

import { People } from '../People';

import { Story } from './People.stories';

const meta: Meta<PageDecoratorArgs> = {
  title: 'Pages/People/Add',
  component: People,
  decorators: [PageDecorator],
  args: { routePath: AppPath.PeoplePage },
  parameters: {
    docs: { story: 'inline', iframeHeight: '500px' },
    msw: graphqlMocks,
  },
};

export default meta;

export const AddNewPerson: Story = {
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);

    await sleep(2000);

    const rowsBeforeAdd = canvas.getAllByRole('row');

    const addButton = canvas.getByRole('button', { name: 'Add' });

    userEvent.click(addButton);

    await sleep(1000);

    const rowsAfterAdd = canvas.getAllByRole('row');

    const firstRow = rowsAfterAdd[1];
    const cells = within(firstRow).getAllByRole('cell');

    expect(cells[1].textContent).toBe('');

    expect(rowsAfterAdd).toHaveLength(rowsBeforeAdd.length + 1);
  },
};
