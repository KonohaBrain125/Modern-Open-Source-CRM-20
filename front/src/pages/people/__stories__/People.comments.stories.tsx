import { expect } from '@storybook/jest';
import type { Meta } from '@storybook/react';
import { userEvent, within } from '@storybook/testing-library';

import { graphqlMocks } from '~/testing/graphqlMocks';

import { People } from '../People';

import { Story } from './People.stories';
import { render } from './shared';

const meta: Meta<typeof People> = {
  title: 'Pages/People/Comments',
  component: People,
};

export default meta;

export const OpenCommentsSection: Story = {
  render,
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);

    const firstRow = await canvas.findByTestId('row-id-1');

    expect(firstRow).toBeDefined();

    const commentsChip = await within(firstRow).findByTestId('comment-chip');
    expect(commentsChip).toBeDefined();

    userEvent.click(commentsChip);
    const commentSection = await canvas.findByText('Comments');
    expect(commentSection).toBeDefined();
  },
  parameters: {
    msw: graphqlMocks,
  },
};
