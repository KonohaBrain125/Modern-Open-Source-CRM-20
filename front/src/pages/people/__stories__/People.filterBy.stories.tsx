import { expect } from '@storybook/jest';
import type { Meta } from '@storybook/react';
import { userEvent, within } from '@storybook/testing-library';
import assert from 'assert';

import {
  PageDecorator,
  type PageDecoratorArgs,
} from '~/testing/decorators/PageDecorator';
import { graphqlMocks } from '~/testing/graphqlMocks';
import { sleep } from '~/testing/sleep';

import { People } from '../People';

import { Story } from './People.stories';

const meta: Meta<PageDecoratorArgs> = {
  title: 'Pages/People/FilterBy',
  component: People,
  decorators: [PageDecorator],
  args: { currentPath: '/people' },
  parameters: {
    docs: { story: 'inline', iframeHeight: '500px' },
    msw: graphqlMocks,
  },
};

export default meta;

export const Email: Story = {
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);

    const filterButton = await canvas.findByText('Filter');
    await userEvent.click(filterButton);

    const emailFilterButton = canvas
      .getAllByTestId('dropdown-menu-item')
      .find((item) => {
        return item.textContent?.includes('Email');
      });

    assert(emailFilterButton);

    await userEvent.click(emailFilterButton);

    const emailInput = canvas.getByPlaceholderText('Email');
    await userEvent.type(emailInput, 'al', {
      delay: 200,
    });

    await sleep(1000);

    expect(await canvas.findByText('Alexandre Prot')).toBeInTheDocument();
    await expect(canvas.queryAllByText('John Doe')).toStrictEqual([]);

    expect(await canvas.findByText('Email:')).toBeInTheDocument();
    expect(await canvas.findByText('Contains al')).toBeInTheDocument();
  },
};

export const CompanyName: Story = {
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);

    const filterButton = await canvas.findByText('Filter');
    await userEvent.click(filterButton);

    const companyFilterButton = canvas
      .getAllByTestId('dropdown-menu-item')
      .find((item) => {
        return item.textContent?.includes('Company');
      });

    assert(companyFilterButton);

    await userEvent.click(companyFilterButton);

    const companyNameInput = canvas.getByPlaceholderText('Company');
    await userEvent.type(companyNameInput, 'Qon', {
      delay: 200,
    });

    await sleep(500);

    const qontoChip = canvas
      .getAllByTestId('dropdown-menu-item')
      .find((item) => {
        return item.textContent?.includes('Qonto');
      });

    expect(qontoChip).toBeInTheDocument();

    assert(qontoChip);

    await userEvent.click(qontoChip);

    // TODO: fix msw where clauses
    // expect(await canvas.findByText('Alexandre Prot')).toBeInTheDocument();
    // await expect(canvas.queryAllByText('John Doe')).toStrictEqual([]);

    expect(await canvas.findByText('Company:')).toBeInTheDocument();
    expect(await canvas.findByText('Is Qonto')).toBeInTheDocument();
  },
};
