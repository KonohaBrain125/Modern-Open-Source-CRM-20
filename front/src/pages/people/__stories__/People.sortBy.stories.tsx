import { expect } from '@storybook/jest';
import type { Meta } from '@storybook/react';
import { userEvent, within } from '@storybook/testing-library';

import {
  PageDecorator,
  type PageDecoratorArgs,
} from '~/testing/decorators/PageDecorator';
import { graphqlMocks } from '~/testing/graphqlMocks';
import { sleep } from '~/testing/sleep';

import { People } from '../People';

import { Story } from './People.stories';

const meta: Meta<PageDecoratorArgs> = {
  title: 'Pages/People/SortBy',
  component: People,
  decorators: [PageDecorator],
  args: { currentPath: '/people' },
  parameters: {
    docs: { story: 'inline', iframeHeight: '500px' },
    msw: graphqlMocks,
  },
};

export default meta;

export const Email: Story = {
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);

    const sortButton = await canvas.findByText('Sort');
    await userEvent.click(sortButton);

    const emailSortButton = canvas.getByText('Email', { selector: 'li > div' });
    await userEvent.click(emailSortButton);

    expect(await canvas.getByTestId('remove-icon-email')).toBeInTheDocument();

    expect(await canvas.findByText('Alexandre Prot')).toBeInTheDocument();
  },
};

export const Cancel: Story = {
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);

    const sortButton = await canvas.findByText('Sort');
    await userEvent.click(sortButton);

    const emailSortButton = canvas.getByText('Email', { selector: 'li > div' });
    await userEvent.click(emailSortButton);

    expect(await canvas.getByTestId('remove-icon-email')).toBeInTheDocument();

    const cancelButton = canvas.getByText('Cancel');
    await userEvent.click(cancelButton);

    await sleep(1000);

    await expect(canvas.queryAllByTestId('remove-icon-email')).toStrictEqual(
      [],
    );
  },
};
