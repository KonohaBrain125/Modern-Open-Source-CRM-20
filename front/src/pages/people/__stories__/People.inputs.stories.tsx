import { getOperationName } from '@apollo/client/utilities';
import { expect } from '@storybook/jest';
import { Meta } from '@storybook/react';
import { userEvent, within } from '@storybook/testing-library';
import { graphql } from 'msw';

import { UPDATE_ONE_PERSON } from '@/people/graphql/mutations/updateOnePerson';
import { SEARCH_COMPANY_QUERY } from '@/search/graphql/queries/searchCompanyQuery';
import { AppPath } from '@/types/AppPath';
import { Company } from '~/generated/graphql';
import {
  PageDecorator,
  PageDecoratorArgs,
} from '~/testing/decorators/PageDecorator';
import { graphqlMocks } from '~/testing/graphqlMocks';
import { fetchOneFromData } from '~/testing/mock-data';
import { mockedCompaniesData } from '~/testing/mock-data/companies';
import { mockedPeopleData } from '~/testing/mock-data/people';
import { sleep } from '~/testing/sleep';

import { People } from '../People';

import { Story } from './People.stories';

const meta: Meta<PageDecoratorArgs> = {
  title: 'Pages/People/Input',
  component: People,
  decorators: [PageDecorator],
  args: { routePath: AppPath.PeoplePage },
  parameters: {
    msw: graphqlMocks,
  },
};

export default meta;

export const InteractWithManyRows: Story = {
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);

    const firstRowEmailCell = await canvas.findByText(
      mockedPeopleData[0].email,
    );

    const secondRowEmailCell = await canvas.findByText(
      mockedPeopleData[1].email,
    );

    expect(
      canvas.queryByTestId('editable-cell-edit-mode-container'),
    ).toBeNull();

    if (!firstRowEmailCell.parentElement) {
      throw new Error('No parent node');
    }

    await userEvent.click(firstRowEmailCell.parentElement);

    expect(
      canvas.queryByTestId('editable-cell-edit-mode-container'),
    ).toBeInTheDocument();

    if (!secondRowEmailCell.parentElement) {
      throw new Error('No parent node');
    }

    await userEvent.click(secondRowEmailCell.parentElement);

    await sleep(25);

    expect(
      canvas.queryByTestId('editable-cell-edit-mode-container'),
    ).toBeNull();

    await userEvent.click(secondRowEmailCell.parentElement);

    await sleep(25);

    expect(
      canvas.queryByTestId('editable-cell-edit-mode-container'),
    ).toBeInTheDocument();
  },
};

export const CheckCheckboxes: Story = {
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);

    await canvas.findByText(mockedPeopleData[0].email);

    const inputCheckboxContainers = await canvas.findAllByTestId(
      'input-checkbox',
    );

    const inputCheckboxes = await canvas.findAllByTestId('input-checkbox');

    const secondCheckboxContainer = inputCheckboxContainers[1];
    const secondCheckbox = inputCheckboxes[1] as HTMLInputElement;

    expect(secondCheckboxContainer).toBeDefined();

    await userEvent.click(secondCheckboxContainer);

    expect(secondCheckbox.checked).toBe(true);

    await userEvent.click(secondCheckbox);

    expect(secondCheckbox.checked).toBe(false);
  },
};

const editRelationMocks = (
  initiallySelectedCompanyName: string,
  searchCompanyNames: Array<string>,
  updateSelectedCompany: Pick<Company, 'name' | 'domainName'>,
) => [
  ...graphqlMocks.filter((graphqlMock) => {
    if (
      typeof graphqlMock.info.operationName === 'string' &&
      [
        getOperationName(UPDATE_ONE_PERSON),
        getOperationName(SEARCH_COMPANY_QUERY),
      ].includes(graphqlMock.info.operationName)
    ) {
      return false;
    }
    return true;
  }),
  ...[
    graphql.mutation(
      getOperationName(UPDATE_ONE_PERSON) ?? '',
      (req, res, ctx) => {
        return res(
          ctx.data({
            updateOnePerson: {
              ...fetchOneFromData(mockedPeopleData, req.variables.where.id),
              ...{
                company: {
                  id: req.variables.where.id,
                  name: updateSelectedCompany.name,
                  domainName: updateSelectedCompany.domainName,
                  __typename: 'Company',
                },
              },
            },
          }),
        );
      },
    ),
    graphql.query(
      getOperationName(SEARCH_COMPANY_QUERY) ?? '',
      (req, res, ctx) => {
        if (!req.variables.where?.AND) {
          // Selected company case
          const searchResults = mockedCompaniesData.filter((company) =>
            [initiallySelectedCompanyName].includes(company.name),
          );
          return res(
            ctx.data({
              searchResults: searchResults,
            }),
          );
        }

        if (
          req.variables.where?.AND?.some(
            (where: { id?: { in: Array<string> } }) => where.id?.in,
          )
        ) {
          // Selected company case
          const searchResults = mockedCompaniesData.filter((company) =>
            [initiallySelectedCompanyName].includes(company.name),
          );
          return res(
            ctx.data({
              searchResults: searchResults,
            }),
          );
        } else {
          // Search case

          const searchResults = mockedCompaniesData.filter((company) =>
            searchCompanyNames.includes(company.name),
          );
          return res(
            ctx.data({
              searchResults: searchResults,
            }),
          );
        }
      },
    ),
  ],
];

export const EditRelation: Story = {
  play: async ({ canvasElement, step }) => {
    const canvas = within(canvasElement);

    await step('Click on second row company cell', async () => {
      const secondRowCompanyCell = await canvas.findByText(
        mockedPeopleData[2].company.name,
      );

      await userEvent.click(
        secondRowCompanyCell.parentNode?.parentNode?.parentNode
          ?.parentElement as HTMLElement,
      );
    });

    await step('Type "Air" in relation picker', async () => {
      const relationInput = await canvas.findByPlaceholderText('Search');

      await userEvent.type(relationInput, 'Air', {
        delay: 200,
      });
    });

    await step('Select "Airbnb"', async () => {
      const airbnbChip = await canvas.findByText('Airbnb', {
        selector: 'div',
      });

      await userEvent.click(airbnbChip);
    });

    await step('Check if Airbnb is in second row company cell', async () => {
      await canvas.findByText('Airbnb');
    });
  },
  parameters: {
    msw: editRelationMocks('Qonto', ['Airbnb', 'Aircall'], {
      name: 'Airbnb',
      domainName: 'airbnb.com',
    }),
  },
};

export const SelectRelationWithKeys: Story = {
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);

    let firstRowCompanyCell = await canvas.findByText(
      mockedPeopleData[0].company.name,
    );
    await sleep(25);

    await userEvent.click(
      firstRowCompanyCell.parentNode?.parentNode?.parentNode
        ?.parentElement as HTMLElement,
    );
    firstRowCompanyCell = await canvas.findByText(
      mockedPeopleData[0].company.name,
    );
    await sleep(25);
    await userEvent.click(firstRowCompanyCell);

    const relationInput = await canvas.findByPlaceholderText('Search');

    await userEvent.type(relationInput, 'Air', {
      delay: 200,
    });

    await userEvent.type(relationInput, '{arrowdown}');

    await sleep(50);

    await userEvent.type(relationInput, '{arrowup}');

    await sleep(50);

    await userEvent.type(relationInput, '{arrowdown}');

    await sleep(50);

    await userEvent.type(relationInput, '{arrowdown}');

    await sleep(50);

    await userEvent.type(relationInput, '{enter}');

    await sleep(200);

    const allAirbns = await canvas.findAllByText('Aircall');
    expect(allAirbns.length).toBe(1);
  },
  parameters: {
    msw: editRelationMocks('Qonto', ['Airbnb', 'Aircall'], {
      name: 'Aircall',
      domainName: 'aircall.io',
    }),
  },
};
