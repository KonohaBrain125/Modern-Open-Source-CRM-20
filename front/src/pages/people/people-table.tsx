import {
  FaRegBuilding,
  FaCalendar,
  FaEnvelope,
  FaRegUser,
  FaMapPin,
  FaPhone,
  FaStream,
  FaUser,
  FaBuilding,
} from 'react-icons/fa';
import { createColumnHelper } from '@tanstack/react-table';
import ClickableCell from '../../components/table/ClickableCell';
import ColumnHead from '../../components/table/ColumnHead';
import Checkbox from '../../components/form/Checkbox';
import CompanyChip from '../../components/chips/CompanyChip';
import { GraphqlQueryPerson, Person } from '../../interfaces/person.interface';
import PipeChip from '../../components/chips/PipeChip';
import EditableText from '../../components/table/editable-cell/EditableText';
import { OrderByFields, updatePerson } from '../../services/people';
import {
  FilterType,
  SortType,
} from '../../components/table/table-header/interface';
import { People_Bool_Exp } from '../../generated/graphql';
import {
  SEARCH_COMPANY_QUERY,
  SEARCH_PEOPLE_QUERY,
} from '../../services/search/search';
import { GraphqlQueryCompany } from '../../interfaces/company.interface';
import EditablePhone from '../../components/table/editable-cell/EditablePhone';
import EditableFullName from '../../components/table/editable-cell/EditableFullName';

export const availableSorts = [
  {
    key: 'fullname',
    label: 'People',
    icon: <FaRegUser />,
  },
  {
    key: 'company_name',
    label: 'Company',
    icon: <FaRegBuilding />,
  },
  {
    key: 'email',
    label: 'Email',
    icon: <FaEnvelope />,
  },
  { key: 'phone', label: 'Phone', icon: <FaPhone /> },
  {
    key: 'created_at',
    label: 'Created at',
    icon: <FaCalendar />,
  },
  { key: 'city', label: 'City', icon: <FaMapPin /> },
] satisfies Array<SortType<OrderByFields>>;

const fullnameFilter = {
  key: 'fullname',
  label: 'People',
  icon: <FaUser />,
  whereTemplate: (operand, { firstname, lastname }) => {
    if (operand.keyWord === 'equal') {
      return {
        _and: [
          { firstname: { _eq: `${firstname}` } },
          { lastname: { _eq: `${lastname}` } },
        ],
      };
    }

    if (operand.keyWord === 'not_equal') {
      return {
        _not: {
          _and: [
            { firstname: { _eq: `${firstname}` } },
            { lastname: { _eq: `${lastname}` } },
          ],
        },
      };
    }
    console.error(Error(`Unhandled operand: ${operand.keyWord}`));
    return {};
  },
  searchQuery: SEARCH_PEOPLE_QUERY,
  searchTemplate: (searchInput: string) => ({
    _or: [
      { firstname: { _ilike: `%${searchInput}%` } },
      { lastname: { _ilike: `%${searchInput}%` } },
    ],
  }),
  searchResultMapper: (person: GraphqlQueryPerson) => ({
    displayValue: `${person.firstname} ${person.lastname}`,
    value: { firstname: person.firstname, lastname: person.lastname },
  }),
  operands: [
    { label: 'Equal', id: 'equal', keyWord: 'equal' },
    { label: 'Not equal', id: 'not-equal', keyWord: 'not_equal' },
  ],
} satisfies FilterType<People_Bool_Exp>;

const companyFilter = {
  key: 'company_name',
  label: 'Company',
  icon: <FaBuilding />,
  whereTemplate: (operand, { companyName }) => {
    if (operand.keyWord === 'equal') {
      return {
        company: { name: { _eq: companyName } },
      };
    }

    if (operand.keyWord === 'not_equal') {
      return {
        _not: { company: { name: { _eq: companyName } } },
      };
    }
    console.error(Error(`Unhandled operand: ${operand.keyWord}`));
    return {};
  },
  searchQuery: SEARCH_COMPANY_QUERY,
  searchTemplate: (searchInput: string) => ({
    name: { _ilike: `%${searchInput}%` },
  }),
  searchResultMapper: (company: GraphqlQueryCompany) => ({
    displayValue: company.name,
    value: { companyName: company.name },
  }),
  operands: [
    { label: 'Equal', id: 'equal', keyWord: 'equal' },
    { label: 'Not equal', id: 'not-equal', keyWord: 'not_equal' },
  ],
} satisfies FilterType<People_Bool_Exp>;

const emailFilter = {
  key: 'email',
  label: 'Email',
  icon: <FaEnvelope />,
  whereTemplate: (operand, { email }) => {
    if (operand.keyWord === 'equal') {
      return {
        email: { _eq: email },
      };
    }

    if (operand.keyWord === 'not_equal') {
      return {
        _not: { email: { _eq: email } },
      };
    }
    console.error(Error(`Unhandled operand: ${operand.keyWord}`));
    return {};
  },
  searchQuery: SEARCH_PEOPLE_QUERY,
  searchTemplate: (searchInput: string) => ({
    email: { _ilike: `%${searchInput}%` },
  }),
  searchResultMapper: (person: GraphqlQueryPerson) => ({
    displayValue: person.email,
    value: { email: person.email },
  }),
  operands: [
    { label: 'Equal', id: 'equal', keyWord: 'equal' },
    { label: 'Not equal', id: 'not-equal', keyWord: 'not_equal' },
  ],
} satisfies FilterType<People_Bool_Exp>;

const cityFilter = {
  key: 'city',
  label: 'City',
  icon: <FaMapPin />,
  whereTemplate: (operand, { city }) => {
    if (operand.keyWord === 'equal') {
      return {
        city: { _eq: city },
      };
    }

    if (operand.keyWord === 'not_equal') {
      return {
        _not: { city: { _eq: city } },
      };
    }
    console.error(Error(`Unhandled operand: ${operand.keyWord}`));
    return {};
  },
  searchQuery: SEARCH_PEOPLE_QUERY,
  searchTemplate: (searchInput: string) => ({
    city: { _ilike: `%${searchInput}%` },
  }),
  searchResultMapper: (person: GraphqlQueryPerson) => ({
    displayValue: person.city,
    value: { city: person.city },
  }),
  operands: [
    { label: 'Equal', id: 'equal', keyWord: 'equal' },
    { label: 'Not equal', id: 'not-equal', keyWord: 'not_equal' },
  ],
} satisfies FilterType<People_Bool_Exp>;

export const availableFilters = [
  fullnameFilter,
  companyFilter,
  emailFilter,
  cityFilter,
  // {
  //   key: 'phone',
  //   label: 'Phone',
  //   icon: faPhone,
  //   whereTemplate: () => ({ phone: { _ilike: '%value%' } }),
  //   searchQuery: GET_PEOPLE,
  //   searchTemplate: { phone: { _ilike: '%value%' } },
  // },
  // {
  //   key: 'created_at',
  //   label: 'Created at',
  //   icon: faCalendar,
  //   whereTemplate: () => ({ created_at: { _eq: '%value%' } }),
  //   searchQuery: GET_PEOPLE,
  //   searchTemplate: { created_at: { _eq: '%value%' } },
  // },
] satisfies FilterType<People_Bool_Exp>[];

const columnHelper = createColumnHelper<Person>();
export const peopleColumns = [
  columnHelper.accessor('id', {
    header: () => (
      <Checkbox id={`person-select-all`} name={`person-select-all`} />
    ),
    cell: (props) => (
      <Checkbox
        id={`person-selected-${props.row.original.email}`}
        name={`person-selected-${props.row.original.email}`}
      />
    ),
  }),
  columnHelper.accessor('firstname', {
    header: () => <ColumnHead viewName="People" viewIcon={<FaRegUser />} />,
    cell: (props) => (
      <EditableFullName
        firstname={props.row.original.firstname}
        lastname={props.row.original.lastname}
        changeHandler={(firstName: string, lastName: string) => {
          const person = props.row.original;
          person.firstname = firstName;
          person.lastname = lastName;
          updatePerson(person).catch((error) => console.error(error)); // TODO: handle error
        }}
      />
    ),
  }),
  columnHelper.accessor('email', {
    header: () => <ColumnHead viewName="Email" viewIcon={<FaEnvelope />} />,
    cell: (props) => (
      <EditableText
        placeholder="Email"
        content={props.row.original.email}
        changeHandler={(value: string) => {
          const person = props.row.original;
          person.email = value;
          updatePerson(person).catch((error) => console.error(error)); // TODO: handle error
        }}
      />
    ),
  }),
  columnHelper.accessor('company', {
    header: () => (
      <ColumnHead viewName="Company" viewIcon={<FaRegBuilding />} />
    ),
    cell: (props) => (
      <ClickableCell href="#">
        <CompanyChip
          name={props.row.original.company.name}
          picture={`https://www.google.com/s2/favicons?domain=${props.row.original.company.domain_name}&sz=256`}
        />
      </ClickableCell>
    ),
  }),
  columnHelper.accessor('phone', {
    header: () => <ColumnHead viewName="Phone" viewIcon={<FaPhone />} />,
    cell: (props) => (
      <EditablePhone
        placeholder="Phone"
        value={props.row.original.phone}
        changeHandler={(value: string) => {
          const person = props.row.original;
          person.phone = value;
          updatePerson(person).catch((error) => console.error(error)); // TODO: handle error
        }}
      />
    ),
  }),
  columnHelper.accessor('creationDate', {
    header: () => <ColumnHead viewName="Creation" viewIcon={<FaCalendar />} />,
    cell: (props) => (
      <ClickableCell href="#">
        {new Intl.DateTimeFormat(undefined, {
          month: 'short',
          day: 'numeric',
          year: 'numeric',
        }).format(props.row.original.creationDate)}
      </ClickableCell>
    ),
  }),
  columnHelper.accessor('pipe', {
    header: () => <ColumnHead viewName="Pipe" viewIcon={<FaStream />} />,
    cell: (props) => (
      <ClickableCell href="#">
        <PipeChip opportunity={props.row.original.pipe} />
      </ClickableCell>
    ),
  }),
  columnHelper.accessor('city', {
    header: () => <ColumnHead viewName="City" viewIcon={<FaMapPin />} />,
    cell: (props) => (
      <EditableText
        shouldAlignRight={true}
        placeholder="City"
        content={props.row.original.city}
        changeHandler={(value: string) => {
          const person = props.row.original;
          person.city = value;
          updatePerson(person).catch((error) => console.error(error)); // TODO: handle error
        }}
      />
    ),
  }),
];
