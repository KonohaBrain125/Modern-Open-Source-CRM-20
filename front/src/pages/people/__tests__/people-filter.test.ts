import {
  assertFilterUseCompanySearch,
  assertFilterUsePeopleSearch,
} from '../../../components/table/table-header/interface';
import { GraphqlQueryPerson } from '../../../interfaces/person.interface';
import { mockCompanyData } from '../../companies/__stories__/mock-data';
import { defaultData } from '../default-data';
import { availableFilters } from '../people-table';

const JohnDoeUser = defaultData.find(
  (user) => user.email === 'john@linkedin.com',
) as GraphqlQueryPerson;

describe('PeopleFilter', () => {
  for (const filter of availableFilters) {
    it(`should render the filter ${filter.key}`, () => {
      if (assertFilterUseCompanySearch(filter)) {
        const filterSelectedValue = filter.searchResultMapper(
          mockCompanyData[0],
        );
        for (const operand of filter.operands) {
          expect(
            filter.whereTemplate(operand, filterSelectedValue.value),
          ).toMatchSnapshot();
        }
      }
      if (assertFilterUsePeopleSearch(filter)) {
        const filterSelectedValue = filter.searchResultMapper(JohnDoeUser);
        for (const operand of filter.operands) {
          expect(
            filter.whereTemplate(operand, filterSelectedValue.value),
          ).toMatchSnapshot();
        }
      }
    });
  }
});
