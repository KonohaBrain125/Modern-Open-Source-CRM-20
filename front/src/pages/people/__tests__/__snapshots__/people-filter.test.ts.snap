// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PeopleFilter should render the filter city 1`] = `
Object {
  "city": Object {
    "_eq": "Paris",
  },
}
`;

exports[`PeopleFilter should render the filter city 2`] = `
Object {
  "_not": Object {
    "city": Object {
      "_eq": "Paris",
    },
  },
}
`;

exports[`PeopleFilter should render the filter company_name 1`] = `
Object {
  "company": Object {
    "name": Object {
      "_eq": "ACME",
    },
  },
}
`;

exports[`PeopleFilter should render the filter company_name 2`] = `
Object {
  "_not": Object {
    "company": Object {
      "name": Object {
        "_eq": "ACME",
      },
    },
  },
}
`;

exports[`PeopleFilter should render the filter email 1`] = `
Object {
  "email": Object {
    "_eq": "john@linkedin.com",
  },
}
`;

exports[`PeopleFilter should render the filter email 2`] = `
Object {
  "_not": Object {
    "email": Object {
      "_eq": "john@linkedin.com",
    },
  },
}
`;

exports[`PeopleFilter should render the filter fullname 1`] = `
Object {
  "_and": Array [
    Object {
      "firstname": Object {
        "_eq": "John",
      },
    },
    Object {
      "lastname": Object {
        "_eq": "Doe",
      },
    },
  ],
}
`;

exports[`PeopleFilter should render the filter fullname 2`] = `
Object {
  "_not": Object {
    "_and": Array [
      Object {
        "firstname": Object {
          "_eq": "John",
        },
      },
      Object {
        "lastname": Object {
          "_eq": "Doe",
        },
      },
    ],
  },
}
`;
